### PR TITLE
[Xamarin.Android.Build.Tasks] Use binutils for Aot and LLVM

### PR DESCRIPTION
**Pull request by pjcollins**
It appears that an Android NDK installation is no longer needed when
using Aot with LLVM.  Projects which enable Aot and LLVM will no longer
attempt to use the NDK unless it is explicitly requested by setting
`$(AndroidNdkDirectory)` to a valid NDK path in the project file.

### CODE REVIEW

File: src/Xamarin.Android.Build.Tasks/Tasks/GetAotArguments.cs
Comment (pjcollins) [inline src/Xamarin.Android.Build.Tasks/Tasks/GetAotArguments.cs:300]: @grendello what are your thoughts on this? Do you remember why we needed to pass these extra lib paths from the NDK, and do you think they would still be needed if we use the `ld` that we now bundle and ship?

I tested template project with AOT+LLVM on a few devices and emulators (including an armeabi Nexux 5 running Android 6.0), and I haven't hit any issues yet...

Comment (grendello) [reply src/Xamarin.Android.Build.Tasks/Tasks/GetAotArguments.cs:300]: As far as I recall, the extra lib paths were for NDK versions which used `gcc`, as we needed to link private gcc libraries.  For the LLVM-based NDK we probably no longer need it, and our native toolchain doesn't need it too.  The only reason to keep them, I think, is to keep support for old NDK versions installed on user machines (we had an instance when somebody still used NDK r16). Other than that, I **think** they can be safely removed.

Comment (pjcollins) [reply src/Xamarin.Android.Build.Tasks/Tasks/GetAotArguments.cs:300]: Should we even continue to allow usage of an arbitrary Android NDK for our AOT tasks if our installers now contain all of the required tools?

Comment (grendello) [reply src/Xamarin.Android.Build.Tasks/Tasks/GetAotArguments.cs:300]: We have no influence on what version of NDK is installed on the user's machine, so we should support as many of them as possible

Comment (pjcollins) [reply src/Xamarin.Android.Build.Tasks/Tasks/GetAotArguments.cs:300]: I guess that's my question, is there a strong reason to allow using an external NDK instead of our toolchain at this point in time?

Comment (grendello) [reply src/Xamarin.Android.Build.Tasks/Tasks/GetAotArguments.cs:300]: AOT tasks should work just fine with our toolchain, it would be good to make that non-optional. Otherwise, we need to keep allowing NDK because of projects which include native code that requires the full C/C++ toolchain.

Comment (jonpryor) [reply src/Xamarin.Android.Build.Tasks/Tasks/GetAotArguments.cs:300]: Rationale for why we may need to support older NDK tooling, e.g. https://github.com/xamarin/xamarin-android/issues/6454

Comment (grendello) [reply src/Xamarin.Android.Build.Tasks/Tasks/GetAotArguments.cs:300]: By default, yes, we should use our embedded toolchain, ignoring `$(AndroidNdkDirectory)`. But we must make it possible for people to use their NDK if they need to because of compiler bugs, toolchain version requirements (for instance for reproducible builds). I can't find another issue we had when someone had to use NDK r16b because newer versions broke some code for them. For these reasons, even though I'd personally prefer to rely only on our embedded toolchain, we need to keep external NDK support in some form.

**Review by grendello — APPROVED** [review]

**Review by jonathanpeppers — APPROVED** [review]
After we merge this, is it worth making the lookup of the NDK in xamarin-android-tools optional?

I haven't measured, but surely some amount of milliseconds probing for NDKs, and we could skip that now?

**Comment by jonathanpeppers** [comment]
Oh, I see the `do-not-merge` label now. Is there an issue?

**Comment by pjcollins** [comment]
> Oh, I see the do-not-merge label now. Is there an issue?

I think there are still a couple of test failures using the latest Android NDK when LLVM is disabled, but I wanted to open this up for review on the general approach of the changes. I'm wondering if we should only allow AOT to use the Android NDK when LLVM is enabled, as that is how things work today?

> After we merge this, is it worth making the lookup of the NDK in xamarin-android-tools optional?

I'm not sure we want to do this quite yet, as `$(_AndroidNdkDirectory)` is still used by `BuildApk` (required when `$(_AndroidCheckedBuild)` is true), and `MakeBundleNativeCodeExternal`.

**Comment by jonathanpeppers** [comment]
Right now I don't think you can use LLVM without having an NDK -- but I was thinking we could fix that? Like it should be possible to work?

If you didn't need the NDK, we could enable LLVM by default for `Release` builds to improve startup times.

**Comment by pjcollins** [comment]
The latest test failures appear to be ignorable. I spent some more time testing this today but am blocked on LLVM configurations with .NET 7 due to a known crash. We may want to consider holding this until we can further test LLVM with .NET 7.

> Right now I don't think you can use LLVM without having an NDK -- but I was thinking we could fix that? Like it should be possible to work?

@jonathanpeppers do you happen to remember a scenario where LLVM relied on the NDK? When previously testing this against .NET 6 and classic LLVM was working for me without the NDK at least in some simple project template cases.

**Comment by pjcollins** [comment]
There are a couple of bugs to revisit in the future once we have the latest 7.0.100 RC builds from runtime, but this should be good to go if CI looks good.

File: src/Xamarin.Android.Build.Tasks/Tasks/GetAotArguments.cs
Comment (grendello) [inline src/Xamarin.Android.Build.Tasks/Tasks/GetAotArguments.cs:310]: I wonder if (at some point, not necessarily in this PR) we could detect the version of LLVM in the NDK and append `--no-relax` if that version is >= 14

**Review by grendello — APPROVED** [review]

**Comment by TomSoPolaris** [comment]
Hi, we're hoping this fixes https://github.com/xamarin/xamarin-android/issues/7352. Has this been released yet? I'm not good at knowing how issues make it into releases and get shipped. Would appreciate some insight.
We had to disable LLVM on our Android app store builds or else they'd insta-crash on launch.

**Pull request merged**
